libs/http2-manager/test/Test/HTTP2/Client/ManagerSpec.hs

```diff
@@ -0,0 +1,325 @@
+    port <- withTestServer mCtx $ \TestServer {..} -> do
+      echoTest mgr (isJust mCtx) serverPort
+      readIORef acceptedConns `shouldReturn` 1
+      pure serverPort
+
+    -- See "should fail with appropriate error when a dead connection is used"
+    -- to know what happens when we don't wait for the background thread to go
+    -- away.
+    Just deadConn <- Map.lookup (isJust mCtx, "localhost", port) <$> readTVarIO (connections mgr)
+    wait $ backgroundThread deadConn
+
+    withTestServerOnPort mCtx port $ \TestServer {..} -> do
+      echoTest mgr (isJust mCtx) port
+      -- this is still 1 because we have a new 'TestServer'
+      readIORef acceptedConns `shouldReturn` 1
+
+data TestException = TestException
+  deriving (Show, Eq)
+
+instance Exception TestException
+
+data TestServer = TestServer
+  { serverThread :: Async (),
+    acceptedConns :: IORef Int,
+    liveConns :: IORef (Map Unique (Async ())),
+    serverPort :: Int
+  }
+
+loadServerSSLContext :: IO SSL.SSLContext
+loadServerSSLContext = do
+  ctx <- SSL.context
+  SSL.contextSetCertificateFile ctx "test/resources/localhost.pem"
+  SSL.contextSetPrivateKeyFile ctx "test/resources/localhost-key.pem"
+  SSL.contextSetALPNProtos ctx ["h2"]
+  SSL.contextSetCiphers ctx "HIGH"
+  sslCheck <- SSL.contextCheckPrivateKey ctx
+  sslCheck `shouldBe` True
+  pure ctx
+
+loadWrongServerSSLContext :: IO SSL.SSLContext
+loadWrongServerSSLContext = do
+  ctx <- SSL.context
+  SSL.contextSetCertificateFile ctx "test/resources/localhost.example.com.pem"
+  SSL.contextSetPrivateKeyFile ctx "test/resources/localhost.example.com-key.pem"
+  SSL.contextSetALPNProtos ctx ["h2"]
+  SSL.contextSetCiphers ctx "HIGH"
+  sslCheck <- SSL.contextCheckPrivateKey ctx
+  sslCheck `shouldBe` True
+  pure ctx
+
+mkTestManager :: IO HTTP2Manager
+mkTestManager = do
+  mgr <- defaultHTTP2Manager
+  SSL.contextSetCAFile (sslContext mgr) "test/resources/unit-ca.pem"
+  pure mgr
+
+withTestServer :: Maybe SSL.SSLContext -> (TestServer -> IO a) -> IO a
+withTestServer mCtx action = do
+  bracket (bindRandomPortTCP "*") (close . snd) $ \(port, sock) ->
+    withTestServerOnSocket mCtx action (port, sock)
+
+withTestServerOnPort :: Maybe SSL.SSLContext -> Int -> (TestServer -> IO a) -> IO a
+withTestServerOnPort mCtx serverPort action = do
+  bracket (bindPortTCP serverPort "*") close $ \listenSock ->
+    withTestServerOnSocket mCtx action (serverPort, listenSock)
+
+withTestServerOnSocket :: Maybe SSL.SSLContext -> (TestServer -> IO a) -> (Int, Socket) -> IO a
+withTestServerOnSocket mCtx action (serverPort, listenSock) = do
+  acceptedConns <- newIORef 0
+  liveConns <- newIORef mempty
+  let cleanupServer serverThread = do
+        cancel serverThread
+        mapM_ cancel =<< readIORef liveConns
+  bracket (async $ testServerOnSocket mCtx listenSock acceptedConns liveConns) cleanupServer $ \serverThread ->
+    action TestServer {..}
+
+allocServerConfig :: Either Socket SSL.SSL -> IO Server.Config
```

Comment:
could this be useful for library users, ie. should it be exported from at least the internals?